File: VirtualWorld-Cpp/Virtual-World/Roslina.cpp
#include "Roslina.h"
#include <iostream>

Roslina::Roslina()
{
}

Roslina::Roslina(Organizm org)
{
}

void Roslina::akcja(std::vector<int>& blackList, std::vector<Organizm*>& sortedList, std::vector <std::vector <Organizm*> >& organizmy)
{

	if (isRadioactive()) {
		getSwiat()->usunWokol(getPolozenie(0), getPolozenie(1), organizmy, this);
	}

	for (int i = 0; i < this->iloscRozp(); i++) {

		int temp = rand() % 4;

		if (!temp) {

			Organizm* uderzajacy = this;

			std::vector<int> teoretyczna{ 0,0 };

			bool czySieWykonalo = 0;

			if (getPolozenie(1) - 1 >= 0 && getSwiat()->sprawdzPole(getPolozenie(0), getPolozenie(1) - 1) == ' ') {
				getSwiat()->edytujPlansze(getPolozenie(0), getPolozenie(1) - 1, this->getZnak());
				getSwiat()->addOrganizmByXY(getPolozenie(0), getPolozenie(1) - 1, organizmy, this->getZnak());
				czySieWykonalo = 1;
			}

			else if (getPolozenie(1) + 1 < getSwiat()->getRozmiarY() && getSwiat()->sprawdzPole(getPolozenie(0), getPolozenie(1) + 1) == ' ') {
				getSwiat()->edytujPlansze(getPolozenie(0), getPolozenie(1) + 1, this->getZnak());
				getSwiat()->addOrganizmByXY(getPolozenie(0), getPolozenie(1) + 1, organizmy, this->getZnak());
				czySieWykonalo = 1;
			}

			else if (getPolozenie(0) - 1 >= 0 && getSwiat()->sprawdzPole(getPolozenie(0) - 1, getPolozenie(1)) == ' ') {
				getSwiat()->edytujPlansze(getPolozenie(0) - 1, getPolozenie(1), this->getZnak());
				getSwiat()->addOrganizmByXY(getPolozenie(0) - 1, getPolozenie(1), organizmy, this->getZnak());
				czySieWykonalo = 1;
			}

			else if (getPolozenie(0) + 1 < getSwiat()->getRozmiarX() && getSwiat()->sprawdzPole(getPolozenie(0) + 1, getPolozenie(1) - 1) == ' ') {
				getSwiat()->edytujPlansze(getPolozenie(0) + 1, getPolozenie(1), this->getZnak());
				getSwiat()->addOrganizmByXY(getPolozenie(0) + 1, getPolozenie(1), organizmy, this->getZnak());
				czySieWykonalo = 1;
			}

			if (czySieWykonalo) {

				std::string str{ "Roslina rozprzestrzenila sie! (" + std::string(1, this->getZnak()) + ")" };

				getSwiat()->addToLog(str);
			}

		}

	}

}

void Roslina::kolizja(std::vector<int> teoretyczna, std::vector<int>& blackList, std::vector<Organizm*>& sortedList, Organizm* uderzony, Organizm* uderzajacy, std::vector <std::vector <Organizm*>>& organizmy)
{

	char lAtakujacego = uderzajacy->getZnak();
	char lAtakowanego = getSwiat()->sprawdzPole(teoretyczna[0], teoretyczna[1]);

	if (czyZabojcza()) {

		getSwiat()->edytujPlansze(uderzajacy->getPolozenie(0), uderzajacy->getPolozenie(1), ' ');
		getSwiat()->deleteOrganizmByXY(uderzajacy->getPolozenie(0), uderzajacy->getPolozenie(1), organizmy);

		getSwiat()->edytujPlansze(teoretyczna[0], teoretyczna[1], ' ');
		getSwiat()->deleteOrganizmByXY(teoretyczna[0], teoretyczna[1], organizmy);

		std::string str{ "Zwierze (" + std::string(1, uderzajacy->getZnak()) + ") zjadlo " + std::string(1, uderzony->getZnak()) + " i nie zyje" };

		getSwiat()->addToLog(str);
	}

	else {

		getSwiat()->edytujPlansze(uderzajacy->getPolozenie(0), uderzajacy->getPolozenie(1), ' ');
		getSwiat()->edytujPlansze(teoretyczna[0], teoretyczna[1], lAtakujacego);

		getSwiat()->deleteOrganizmByXY(teoretyczna[0], teoretyczna[1], organizmy);
		getSwiat()->changePositionOrganizm(uderzajacy->getPolozenie(0), uderzajacy->getPolozenie(1), teoretyczna[0], teoretyczna[1], organizmy);

		getSwiat()->zmienSileByXY(uderzajacy->getPolozenie(0), uderzajacy->getPolozenie(1), organizmy, uderzony->dodawanaSilaPoZjedzeniu());

		if (uderzony->dodawanaSilaPoZjedzeniu()) {
			std::string str{ "Atakujacy (" + std::string(1, uderzajacy->getZnak()) + ") zwiekszyl se sile o " + std::to_string(uderzony->dodawanaSilaPoZjedzeniu()) };

			getSwiat()->addToLog(str);
		}

		std::string str{ "Roslina zostala zjedzona! (" + std::string(1, uderzony->getZnak()) + ") przez (" + std::string(1, uderzajacy->getZnak()) + ")" };

		getSwiat()->addToLog(str);

	}
	
	for (int j = 0; j < sortedList.size(); j++) {

		if (sortedList[j]->getPolozenie(0) == uderzajacy->getPolozenie(0) && sortedList[j]->getPolozenie(1) == uderzajacy->getPolozenie(1))   blackList.push_back(j);
		if (sortedList[j]->getPolozenie(0) == teoretyczna[0] && sortedList[j]->getPolozenie(1) == teoretyczna[1] && !czyZabojcza() )   blackList.push_back(j);

	}

}

void Roslina::rysowanie()
{
}

bool Roslina::isPlant()
{
	return 1;
}

char Roslina::getZnak()
{
	return 0;
}

bool Roslina::czyZabojcza()
{
	return false;
}

int Roslina::dodawanaSilaPoZjedzeniu()
{
	return 0;
}

bool Roslina::isRadioactive()
{
	return false;
}

Roslina::~Roslina()
{
}

int Roslina::iloscRozp()
{
	return 1;
}

File: VirtualWorld-Java/src/Roslina.java
import java.util.ArrayList;
import java.util.Random;

public class Roslina extends Organizm {

	public Roslina(){
	}
	
	public Roslina(Organizm org){
		
	}
	
	@Override
	public void akcja(ArrayList<Integer> blackList, ArrayList<Organizm> sortedList, ArrayList<ArrayList<Organizm>> organizmy) {
		
		if (isRadioactive()) {
			getSwiat().usunWokol(getPolozenie(0), getPolozenie(1), organizmy, this);
		}

		for (int i = 0; i < this.iloscRozp(); i++) {

			Random rand = new Random();
			int temp = rand.nextInt(4);

			if (temp == 0) {

				ArrayList<Integer> teoretyczna = new ArrayList<Integer>();
				teoretyczna.add(0); teoretyczna.add(0);
				
				String s = ""+this.getZnak();
				Boolean czySieWykonalo = false;

				if (getPolozenie(1) - 1 >= 0 && getSwiat().sprawdzPole(getPolozenie(0), getPolozenie(1) - 1) == ' ') {
					getSwiat().edytujPlansze(getPolozenie(0), getPolozenie(1) - 1, s);
					getSwiat().addOrganizmByXY(getPolozenie(0), getPolozenie(1) - 1, organizmy, this.getZnak());
					czySieWykonalo = true;
				}

				else if (getPolozenie(1) + 1 < getSwiat().getRozmiarY() && getSwiat().sprawdzPole(getPolozenie(0), getPolozenie(1) + 1) == ' ') {
					getSwiat().edytujPlansze(getPolozenie(0), getPolozenie(1) + 1, s);
					getSwiat().addOrganizmByXY(getPolozenie(0), getPolozenie(1) + 1, organizmy, this.getZnak());
					czySieWykonalo = true;
				}

				else if (getPolozenie(0) - 1 >= 0 && getSwiat().sprawdzPole(getPolozenie(0) - 1, getPolozenie(1)) == ' ') {
					getSwiat().edytujPlansze(getPolozenie(0) - 1, getPolozenie(1), s);
					getSwiat().addOrganizmByXY(getPolozenie(0) - 1, getPolozenie(1), organizmy, this.getZnak());
					czySieWykonalo = true;
				}

				else if (getPolozenie(0) + 1 < getSwiat().getRozmiarX() && getSwiat().sprawdzPole(getPolozenie(0) + 1, getPolozenie(1) - 1) == ' ') {
					getSwiat().edytujPlansze(getPolozenie(0) + 1, getPolozenie(1), s);
					getSwiat().addOrganizmByXY(getPolozenie(0) + 1, getPolozenie(1), organizmy, this.getZnak());
					czySieWykonalo = true;
				}

				if (czySieWykonalo) {

					String str = "Roslina rozprzestrzenila sie! (" + this.getZnak() + ")";

					getSwiat().addToLog(str);
				}

			}

		}
		
	}
	
	@Override
	public void kolizja(ArrayList<Integer> teoretyczna, ArrayList<Integer> blackList, ArrayList<Organizm> sortedList, Organizm uderzony, Organizm uderzajacy, ArrayList<ArrayList<Organizm>> organizmy) {
		
		char lAtakujacego = uderzajacy.getZnak();
		if (czyZabojcza()) {

			getSwiat().edytujPlansze(uderzajacy.getPolozenie(0), uderzajacy.getPolozenie(1), " ");
			getSwiat().deleteOrganizmByXY(uderzajacy.getPolozenie(0), uderzajacy.getPolozenie(1), organizmy);

			getSwiat().edytujPlansze(teoretyczna.get(0), teoretyczna.get(1), " ");
			getSwiat().deleteOrganizmByXY(teoretyczna.get(0), teoretyczna.get(1), organizmy);

			String str = "Zwierze (" + uderzajacy.getZnak() + ") zjadlo " + uderzony.getZnak() + " i nie zyje";

			getSwiat().addToLog(str);
		}

		else {

			getSwiat().edytujPlansze(uderzajacy.getPolozenie(0), uderzajacy.getPolozenie(1), " ");
			String s = ""+lAtakujacego;
			getSwiat().edytujPlansze(teoretyczna.get(0), teoretyczna.get(1), s);

			getSwiat().deleteOrganizmByXY(teoretyczna.get(0), teoretyczna.get(1), organizmy);
			getSwiat().changePositionOrganizm(uderzajacy.getPolozenie(0), uderzajacy.getPolozenie(1), teoretyczna.get(0), teoretyczna.get(1), organizmy);

			getSwiat().zmienSileByXY(uderzajacy.getPolozenie(0), uderzajacy.getPolozenie(1), organizmy, uderzony.dodawanaSilaPoZjedzeniu());

			if (uderzony.dodawanaSilaPoZjedzeniu() > 0) {
				String str = "Atakujacy (" + uderzajacy.getZnak() + ") zwiekszyl se sile o " + Integer.toString(uderzony.dodawanaSilaPoZjedzeniu());

				getSwiat().addToLog(str);
			}

			String str = "Roslina zostala zjedzona! (" + uderzony.getZnak() + ") przez (" + uderzajacy.getZnak() + ")";

			getSwiat().addToLog(str);

		}
		
		for (int j = 0; j < sortedList.size(); j++) {

			if (sortedList.get(j).getPolozenie(0) == uderzajacy.getPolozenie(0) && sortedList.get(j).getPolozenie(1) == uderzajacy.getPolozenie(1))   blackList.add(j);
			if (sortedList.get(j).getPolozenie(0) == teoretyczna.get(0) && sortedList.get(j).getPolozenie(1) == teoretyczna.get(1) && !czyZabojcza() )   blackList.add(j);

		}
		
	}
	
	@Override
	public void rysowanie() {
		System.out.println('R');
	}
	
	@Override
	public char getZnak() {
		return 'R';
	}
	
	@Override
	public Boolean isPlant() {
		return true;
	}
	
	@Override
	public Boolean czyZabojcza()
	{
		return false;
	}

	@Override
	public int dodawanaSilaPoZjedzeniu()
	{
		return 0;
	}

	@Override
	public Boolean isRadioactive()
	{
		return false;
	}

	@Override
	public int iloscRozp()
	{
		return 1;
	}
	
	
}
